sky/serve/autoscalers.py

```diff
@@ -192,49 +234,203 @@ def evaluate_scaling(
+        number of spot instances to provision (including overprovisioning),
+        scale up or down accordingly.
+        """
+        scaling_options = []
+        num_to_provision = (self.target_num_replicas + self.num_overprovision)
+
+        # Scale spot instances.
+        current_considered_zones: List[str] = []
+        if num_launched_spot < num_to_provision:
+            # Not enough spot instances, scale up.
+            # Consult spot_placer for the zone to launch spot instance.
+            num_spot_to_scale_up = num_to_provision - num_launched_spot
+            for _ in range(num_spot_to_scale_up):
+                spot_override = self._get_spot_resources_override_dict()
+                zone = self.spot_placer.select(launched_replica_infos,
+                                               current_considered_zones)
+                current_considered_zones.append(zone)
+                spot_override.update({'zone': zone})
+                logger.info(f'Chosen zone {zone} with {self.spot_placer}')
+                scaling_options.append(
+                    AutoscalerDecision(AutoscalerDecisionOperator.SCALE_UP,
+                                       target=spot_override))
+        elif num_launched_spot > num_to_provision:
+            # Too many spot instances, scale down.
+            # Get the replica to scale down with get_replica_ids_to_scale_down
+            num_spot_to_scale_down = num_launched_spot - num_to_provision
+            for replica_id in (
+                    RequestRateAutoscaler.get_replica_ids_to_scale_down(
+                        num_spot_to_scale_down, [
+                            info for info in launched_replica_infos
+                            if info.is_spot
+                        ])):
+                scaling_options.append(
+                    AutoscalerDecision(AutoscalerDecisionOperator.SCALE_DOWN,
+                                       target=replica_id))
+        return scaling_options
+
+    def evaluate_scaling(
+        self,
+        replica_infos: List['replica_managers.ReplicaInfo'],
+    ) -> List[AutoscalerDecision]:
+
+        launched_replica_infos = [
+            info for info in replica_infos if info.is_launched
+        ]
+        self.target_num_replicas = self._get_desired_num_replicas()
+
+        num_launched_spot, num_ready_spot = 0, 0
+        for info in launched_replica_infos:
+            if info.status == serve_state.ReplicaStatus.READY:
+                num_ready_spot += 1
+            num_launched_spot += 1
+        logger.info(f'Number of alive spot instances: {num_launched_spot}, '
+                    f'Number of ready spot instances: {num_ready_spot}')
+
+        scaling_options = (self._scale_spot_instances(num_launched_spot,
+                                                      launched_replica_infos))
+
+        return scaling_options
+
+
+class SpotOnDemandRequestRateAutoscaler(SpotRequestRateAutoscaler):
```

Comment:
Could we combine SpotRequestRateAutoscaler and SpotOnDemandRequestRateAutoscaler into one autoscaler? IIUC we won't use SpotRequestRateAutoscaler alone?

Reply:
I think separating them is kinda useful if we want to run some evaluations of baseline policies without on-demand fallback?

Reply:
Some users might want to only launch spot instances to maximize cost saving.

Reply:
That makes sense! SG.